src/include/common/numeric_utils.h

```diff
@@ -0,0 +1,85 @@
+#pragma once
+
+#include <type_traits>
+
+#include "common/types/int128_t.h"
+#include "common/types/types.h"
+#include <bit>
+#include <concepts>
+
+namespace kuzu {
+namespace common {
+namespace NumericUtils {
+
+template<typename T>
+concept IsIntegral = std::integral<T> || std::same_as<std::remove_cvref_t<T>, int128_t>;
+
+template<typename T>
+concept IsSigned = std::same_as<T, int128_t> || std::numeric_limits<T>::is_signed;
+
+template<typename T>
+concept IsUnSigned = std::numeric_limits<T>::is_unsigned;
+
+template<typename T>
+struct MakeSigned {
+    using type = std::make_signed_t<T>;
+};
+
+template<>
+struct MakeSigned<int128_t> {
+    using type = int128_t;
+};
+
+template<typename T>
+using MakeSignedT = typename MakeSigned<T>::type;
+
+template<typename T>
+struct MakeUnSigned {
+    using type = std::make_unsigned_t<T>;
+};
+
+template<>
+struct MakeUnSigned<int128_t> {
+    // currently evaluates to int128_t as we don't have an uint128_t type
+    using type = int128_t;
+};
+
+template<typename T>
+using MakeUnSignedT = typename MakeUnSigned<T>::type;
+
+template<typename T>
+decltype(auto) makeValueSigned(T value) {
+    return static_cast<MakeSignedT<T>>(value);
+}
+
+template<typename T>
+decltype(auto) makeValueUnSigned(T value) {
+    return static_cast<MakeUnSignedT<T>>(value);
+}
+
+template<typename T>
+constexpr int bitWidth(T x) {
+    return std::bit_width(x);
+}
+
+template<>
+constexpr int bitWidth<int128_t>(int128_t x) {
+    if (x.high != 0) {
+        constexpr size_t BITS_PER_BYTE = 8;
+        return sizeof(x.low) * BITS_PER_BYTE + std::bit_width(makeValueUnSigned(x.high));
+    }
+    return std::bit_width(x.low);
+}
+
+// produces bitmasks in the form ...111111
+// can produce masks with # of bits up to sizeof(T) * 8 - 1
+template<IsIntegral T>
+inline T getSaturatedBitmask(size_t width) {
+    static constexpr uint8_t numBitsInT = sizeof(T) * 8;
+    const T fullMask = ~(T(1) << (numBitsInT - 1));
+    const size_t numBitsToDiscard = (numBitsInT - 1 - width);
+    return (fullMask >> numBitsToDiscard);
```

Comment:
Yea this was basically a hack to get around the overflow checks in `int128_t`. We don't have an unsigned version and if the shift value is 127 it'll complain about subtracting 1 from `INT128_MIN`.
Not using `~T(0) >> ...` was to avoid the sign bit filling that comes with right shifting :smile: 
I'll think a bit harder about seeing if I can find a cleaner approach I guess worst case I can have this as a specialization for `int128_t` only so it doesn't mess with the primitive integer types.

Reply:
>  Not using ~T(0) >> ... was to avoid the sign bit filling that comes with right shifting 😄

Good point, I wasn't thinking about int128 being signed when I wrote that. If this works, then I guess leave it, though maybe leave a comment indicating that it has to avoid overflows since it needs to work for int128.

This sounds like a good argument for a uint128 type, even just for internal use. Signed overflow may be undefined, but a lot of stuff relies on unsigned overflow being well-defined. That may be more work than it's worth though; I don't think we'll need to do stuff like implementing an int128 version of our hash function.
https://github.com/kuzudb/kuzu/blob/21dfb5b72d26e92a29ee87b8aaff03d2ce34be77/src/include/function/hash/hash_functions.h#L20-L24